### ISSUE TITLE
Thumbnail Images not generated in PHP 8.1.3 with EE 7.3.2 on Windows Server 2016

### ISSUE THREAD

**Issue by rmaclemale** [ExpressionEngine/ExpressionEngine#3554]
**Description of the problem**

We are running PHP 8.1.15 on Windows Server 2016.  I set up a new instance of EE 7.3.2.  Everything works fine except on the Files screen, where file uploading works fine but thumbnails are not working.  On the Windows version of PHP 8, the php_gd2 extension was renamed php_gd.  I read that EE 7 uses gd2 to generate thumbnail images, so this might be related.

Bryan Nielson suggested on Slack that I add this to the config.php file:  $config['image_resize_protocol'] = 'gd';

I added it, and it didn't help.  I also tried both GD and GD 2 as the Protocol under Content & Design > Settings > Image Resizing.  Neither of them worked.  When I upload a file, EE is creating a 0 KB empty image file in the _thumbs folder.  

**How To Reproduce**

1. Install PHP 8.1.15 on Windows Server 2016 (to use with IIS Manager)
2. Do a clean install of EE 7.3.2 (including setting up a database)
3. Add a new folder inside the install named "library" (this is where we will upload an image file)
4. In IIS Manager, configure the new site to use PHP 8.1.15
5. In IIS Manager, ensure that the php_gd.dll extension is enabled
6. Log into the Dashboard of the new site and go to Content > Files and create a new upload directory for the library folder we added in step 3
7. Upload any jpg or png file that is at least 50 KB in size.  The file will upload fine, but the image thumbnail will display a broken image icon
8. Go into the library\thumbs folder.  There will be an image file there for the file you just uploaded, and it will be 0 KB, aka an empty file.

**Error Messages**
Even with Error Visibility set to "Shown to Super Admins only" in Debugging & Output, there is no error message.  Also, the PHP error log shows nothing for these file uploads.

**Screenshots / Videos / Template Code**
Here's what the attempted file upload thumbnails look like:
https://connectplus.pasco.k12.fl.us/web/empty.png)

**Environment Details:**
 - Version: EE 7.3.2
 - PHP Version: 8.1.15
 - MySQL Version: 5.7.23
 - OS: Windows Server 2016
 - Web Server: IIS

**Possible Solution**
I have no ideas on how to solve this, but suspect it has to do with the Windows version of PHP renaming php_gd2 to php_gd.  

**Additional context**
I ran var_dump(gd_info()); in a php file to ensure that the gd extension was functioning, and it is.  Here is the output from that:  
array(15) { [“GD Version”]=> string(26) “bundled (2.1.0 compatible)” [“FreeType Support”]=> bool(true) [“FreeType Linkage”]=> string(13) “with freetype” [“GIF Read Support”]=> bool(true) [“GIF Create Support”]=> bool(true) [“JPEG Support”]=> bool(true) [“PNG Support”]=> bool(true) [“WBMP Support”]=> bool(true) [“XPM Support”]=> bool(true) [“XBM Support”]=> bool(true) [“WebP Support”]=> bool(true) [“BMP Support”]=> bool(true) [“AVIF Support”]=> bool(true) [“TGA Read Support”]=> bool(true) [“JIS-mapped Japanese Font Support”]=> bool(false) }

I don't see anything else I can try on my end, but I am very happy to test out any possible fixes.

**Comment by intoeetive** [comment]
@rmaclemale can you try also making this change: in file `system\ee\legacy\libraries\Image_lib.php` replace the line 466 that reads
> if ($this->image_library == 'gd2' and function_exists('imagecreatetruecolor')) {
with
> if (in_array($this->image_library, ['gd, 'gd2']) and function_exists('imagecreatetruecolor')) {

**Comment by rmaclemale** [comment]
Sorry for the late response - I was on vacation for 2 weeks.  I inserted the above code, but it did not correct the issue - EE is uploading the file, but it's generating a 0 KB thumbnail in the _thumbs folder.  Also, in the above code, I had to add an apostrophe after 'gd  (I'm assuming that was a typo.)

**Comment by intoeetive** [comment]
This is... interesting.

Can you go to Tools -> Untilities -> PHP Info in EE CP and make screenshot of 'gd' section and post it here?

Also perhaps you could do some debugging here.
Before the abovementioned code, add this:
```
dump($this->image_library);
dd(function_exists('imagecreatetruecolor'));
```
Upload a file with Network tab in your browser's developer tools open. The upload will not finish, but the last request shown in Network tab should give some info - please post that here

**Comment by rmaclemale** [comment]
Here's the last request, copied from Inspector/Network:

XMLHttpRequest {onreadystatechange: null, readyState: 4, timeout: 0, withCredentials: false, upload: XMLHttpRequestUpload, …}
onabort
: 
null
onerror
: 
null
onload
: 
null
onloadend
: 
null
onloadstart
: 
null
onprogress
: 
null
onreadystatechange
: 
null
ontimeout
: 
null
readyState
: 
4
response
: 
"{\"messageType\":\"error\",\"message\":\"syntax error, unexpected identifier &quot;gd2&quot;, expecting &quot;]&quot;\",\"trace\":[\"#0 system\\\\ee\\\\legacy\\\\core\\\\Loader.php(160): EE_Loader-&gt;_ci_load_class()\",\"#1 system\\\\ee\\\\legacy\\\\libraries\\\\Filemanager.php(1136): EE_Loader-&gt;library()\",\"#2 system\\\\ee\\\\legacy\\\\libraries\\\\Filemanager.php(390): Filemanager-&gt;create_thumb()\",\"#3 system\\\\ee\\\\legacy\\\\libraries\\\\Filemanager.php(2015): Filemanager-&gt;save_file()\",\"#4 system\\\\ee\\\\legacy\\\\libraries\\\\Filemanager.php(1089): Filemanager-&gt;_upload_file()\",\"#5 system\\\\ee\\\\ExpressionEngine\\\\Service\\\\File\\\\Upload.php(334): Filemanager-&gt;upload_file()\",\"#6 system\\\\ee\\\\ExpressionEngine\\\\Addons\\\\filepicker\\\\mcp.filepicker.php(325): ExpressionEngine\\\\Service\\\\File\\\\Upload-&gt;uploadTo()\",\"#7 [internal function]: Filepicker_mcp-&gt;ajaxUpload()\",\"#8 system\\\\ee\\\\ExpressionEngine\\\\Controller\\\\Addons\\\\Addons.php(1590): call_user_func_array()\",\"#9 system\\\\ee\\\\ExpressionEngine\\\\Controller\\\\Addons\\\\Addons.php(866): ExpressionEngine\\\\Controller\\\\Addons\\\\Addons-&gt;getModuleSettings()\",\"#10 [internal function]: ExpressionEngine\\\\Controller\\\\Addons\\\\Addons-&gt;settings()\",\"#11 system\\\\ee\\\\ExpressionEngine\\\\Core\\\\Core.php(268): call_user_func_array()\",\"#12 system\\\\ee\\\\ExpressionEngine\\\\Core\\\\Core.php(124): ExpressionEngine\\\\Core\\\\Core-&gt;runController()\",\"#13 system\\\\ee\\\\ExpressionEngine\\\\Boot\\\\boot.php(184): ExpressionEngine\\\\Core\\\\Core-&gt;run()\",\"#14 admin.php(139): require_once(&#039;...&#039;)\",\"#15 {main}\"]}"
responseText
: 
"{\"messageType\":\"error\",\"message\":\"syntax error, unexpected identifier &quot;gd2&quot;, expecting &quot;]&quot;\",\"trace\":[\"#0 system\\\\ee\\\\legacy\\\\core\\\\Loader.php(160): EE_Loader-&gt;_ci_load_class()\",\"#1 system\\\\ee\\\\legacy\\\\libraries\\\\Filemanager.php(1136): EE_Loader-&gt;library()\",\"#2 system\\\\ee\\\\legacy\\\\libraries\\\\Filemanager.php(390): Filemanager-&gt;create_thumb()\",\"#3 system\\\\ee\\\\legacy\\\\libraries\\\\Filemanager.php(2015): Filemanager-&gt;save_file()\",\"#4 system\\\\ee\\\\legacy\\\\libraries\\\\Filemanager.php(1089): Filemanager-&gt;_upload_file()\",\"#5 system\\\\ee\\\\ExpressionEngine\\\\Service\\\\File\\\\Upload.php(334): Filemanager-&gt;upload_file()\",\"#6 system\\\\ee\\\\ExpressionEngine\\\\Addons\\\\filepicker\\\\mcp.filepicker.php(325): ExpressionEngine\\\\Service\\\\File\\\\Upload-&gt;uploadTo()\",\"#7 [internal function]: Filepicker_mcp-&gt;ajaxUpload()\",\"#8 system\\\\ee\\\\ExpressionEngine\\\\Controller\\\\Addons\\\\Addons.php(1590): call_user_func_array()\",\"#9 system\\\\ee\\\\ExpressionEngine\\\\Controller\\\\Addons\\\\Addons.php(866): ExpressionEngine\\\\Controller\\\\Addons\\\\Addons-&gt;getModuleSettings()\",\"#10 [internal function]: ExpressionEngine\\\\Controller\\\\Addons\\\\Addons-&gt;settings()\",\"#11 system\\\\ee\\\\ExpressionEngine\\\\Core\\\\Core.php(268): call_user_func_array()\",\"#12 system\\\\ee\\\\ExpressionEngine\\\\Core\\\\Core.php(124): ExpressionEngine\\\\Core\\\\Core-&gt;runController()\",\"#13 system\\\\ee\\\\ExpressionEngine\\\\Boot\\\\boot.php(184): ExpressionEngine\\\\Core\\\\Core-&gt;run()\",\"#14 admin.php(139): require_once(&#039;...&#039;)\",\"#15 {main}\"]}"
responseType
: 
""
responseURL
: 
"http://www.pasco.k12.fl.us.test/admin.php?/cp/addons/settings/filepicker/ajax-upload"
responseXML
: 
null
status
: 
500
statusText
: 
"Internal Server Error"
timeout
: 
0
upload
: 
XMLHttpRequestUpload
onabort
: 
null
onerror
: 
null
onload
: 
null
onloadend
: 
null
onloadstart
: 
null
onprogress
: 
null
ontimeout
: 
null
[[Prototype]]
: 
XMLHttpRequestUpload
constructor
: 
ƒ XMLHttpRequestUpload()
Symbol(Symbol.toStringTag)
: 
"XMLHttpRequestUpload"
onabort
: 
(...)
onerror
: 
(...)
onload
: 
(...)
onloadend
: 
(...)
onloadstart
: 
(...)
onprogress
: 
(...)
ontimeout
: 
(...)
[[Prototype]]
: 
XMLHttpRequestEventTarget
withCredentials
: 
false
[[Prototype]]
: 
XMLHttpRequest
DONE
: 
4
HEADERS_RECEIVED
: 
2
LOADING
: 
3
OPENED
: 
1
UNSENT
: 
0
abort
: 
ƒ abort()
getAllResponseHeaders
: 
ƒ getAllResponseHeaders()
getResponseHeader
: 
ƒ getResponseHeader()
onreadystatechange
: 
(...)
open
: 
ƒ open()
overrideMimeType
: 
ƒ overrideMimeType()
readyState
: 
(...)
response
: 
(...)
responseText
: 
(...)
responseType
: 
(...)
responseURL
: 
(...)
responseXML
: 
(...)
send
: 
ƒ send()
length
: 
0
name
: 
"send"
arguments
: 
(...)
caller
: 
(...)
[[Prototype]]
: 
ƒ ()
[[Scopes]]
: 
Scopes[0]
setRequestHeader
: 
ƒ setRequestHeader()
status
: 
(...)
statusText
: 
(...)
timeout
: 
(...)
upload
: 
(...)
withCredentials
: 
(...)
constructor
: 
ƒ XMLHttpRequest()
Symbol(Symbol.toStringTag)
: 
"XMLHttpRequest"
onabort
: 
(...)
onerror
: 
(...)
onload
: 
(...)
onloadend
: 
(...)
onloadstart
: 
(...)
onprogress
: 
(...)
ontimeout
: 
(...)
get onreadystatechange
: 
ƒ onreadystatechange()
set onreadystatechange
: 
ƒ onreadystatechange()
get readyState
: 
ƒ readyState()
get response
: 
ƒ response()
get responseText
: 
ƒ responseText()
get responseType
: 
ƒ responseType()
set responseType
: 
ƒ responseType()
get responseURL
: 
ƒ responseURL()
get responseXML
: 
ƒ responseXML()
get status
: 
ƒ status()
get statusText
: 
ƒ statusText()
get timeout
: 
ƒ timeout()
set timeout
: 
ƒ timeout()
get upload
: 
ƒ upload()
get withCredentials
: 
ƒ withCredentials()
set withCredentials
: 
ƒ withCredentials()
[[Prototype]]
: 
XMLHttpRequestEventTarget
onabort
: 
(...)
onerror
: 
(...)
onload
: 
(...)
onloadend
: 
(...)
onloadstart
: 
(...)
onprogress
: 
(...)
ontimeout
: 
(...)
constructor
: 
ƒ XMLHttpRequestEventTarget()
Symbol(Symbol.toStringTag)
: 
"XMLHttpRequestEventTarget"
get onabort
: 
ƒ onabort()
set onabort
: 
ƒ onabort()
get onerror
: 
ƒ onerror()
set onerror
: 
ƒ onerror()
get onload
: 
ƒ onload()
set onload
: 
ƒ onload()
get onloadend
: 
ƒ onloadend()
set onloadend
: 
ƒ onloadend()
get onloadstart
: 
ƒ onloadstart()
set onloadstart
: 
ƒ onloadstart()
get onprogress
: 
ƒ onprogress()
set onprogress
: 
ƒ onprogress()
get ontimeout
: 
ƒ ontimeout()
set ontimeout
: 
ƒ ontimeout()
[[Prototype]]
: 
EventTarget

**Comment by intoeetive** [comment]
It looks like there's a syntax error - I think you did not add closing `'` in `['gd, 'gd2']` this time (which should be ` ['gd', 'gd2'])

HOWEVER - I just realized that there have been updates to Filemanager between version 7.3.2 that you are running and most recent version. Would you mind trying to upgrade? IDK if that might solve your issue by itself, as it seem to be quite specific - but it would be good to use latest codebase for sure

**Comment by rmaclemale** [comment]
OK, the test site has been upgraded to 7.3.6, apostrophe added to gd, and dump code added.  I uploaded an image with Inspector/Network open, and here's what it gave me:

Uncaught SyntaxError: Unexpected token '<', "<pre>strin"... is not valid JSON
    at JSON.parse (<anonymous>)
    at XMLHttpRequest.<anonymous> (admin.php?/cp&C=javascript&M=combo_load&ui=droppable,draggable&file=cp/files/picker,fields/file/control_panel,fields/file/file_field_drag_and_drop,cp/publish/entry-list,fields/file/concurrency_queue,fields/file/file_upload_progress_table,fields/file/drag_and_drop_upload,fields/grid/file_grid,cp/date_picker,cp/files/manager,cp/files/copy-url,cp/files/menu&v=1689006136:3263:34)
(anonymous) @ admin.php?/cp&C=javascript&M=combo_load&ui=droppable,draggable&file=cp/files/picker,fields/file/control_panel,fields/file/file_field_drag_and_drop,cp/publish/entry-list,fields/file/concurrency_queue,fields/file/file_upload_progress_table,fields/file/drag_and_drop_upload,fields/grid/file_grid,cp/date_picker,cp/files/manager,cp/files/copy-url,cp/files/menu&v=1689006136:3263
XMLHttpRequest.send (async)
(anonymous) @ admin.php?/cp&C=javascript&M=combo_load&ui=droppable,draggable&file=cp/files/picker,fields/file/control_panel,fields/file/file_field_drag_and_drop,cp/publish/entry-list,fields/file/concurrency_queue,fields/file/file_upload_progress_table,fields/file/drag_and_drop_upload,fields/grid/file_grid,cp/date_picker,cp/files/manager,cp/files/copy-url,cp/files/menu&v=1689006136:3390
makeUploadPromise @ admin.php?/cp&C=javascript&M=combo_load&ui=droppable,draggable&file=cp/files/picker,fields/file/control_panel,fields/file/file_field_drag_and_drop,cp/publish/entry-list,fields/file/concurrency_queue,fields/file/file_upload_progress_table,fields/file/drag_and_drop_upload,fields/grid/file_grid,cp/date_picker,cp/files/manager,cp/files/copy-url,cp/files/menu&v=1689006136:3234
(anonymous) @ admin.php?/cp&C=javascript&M=combo_load&ui=droppable,draggable&file=cp/files/picker,fields/file/control_panel,fields/file/file_field_drag_and_drop,cp/publish/entry-list,fields/file/concurrency_queue,fields/file/file_upload_progress_table,fields/file/drag_and_drop_upload,fields/grid/file_grid,cp/date_picker,cp/files/manager,cp/files/copy-url,cp/files/menu&v=1689006136:2966
start @ admin.php?/cp&C=javascript&M=combo_load&ui=droppable,draggable&file=cp/files/picker,fields/file/control_panel,fields/file/file_field_drag_and_drop,cp/publish/entry-list,fields/file/concurrency_queue,fields/file/file_upload_progress_table,fields/file/drag_and_drop_upload,fields/grid/file_grid,cp/date_picker,cp/files/manager,cp/files/copy-url,cp/files/menu&v=1689006136:2683
enqueue @ admin.php?/cp&C=javascript&M=combo_load&ui=droppable,draggable&file=cp/files/picker,fields/file/control_panel,fields/file/file_field_drag_and_drop,cp/publish/entry-list,fields/file/concurrency_queue,fields/file/file_upload_progress_table,fields/file/drag_and_drop_upload,fields/grid/file_grid,cp/date_picker,cp/files/manager,cp/files/copy-url,cp/files/menu&v=1689006136:2670
(anonymous) @ admin.php?/cp&C=javascript&M=combo_load&ui=droppable,draggable&file=cp/files/picker,fields/file/control_panel,fields/file/file_field_drag_and_drop,cp/publish/entry-list,fields/file/concurrency_queue,fields/file/file_upload_progress_table,fields/file/drag_and_drop_upload,fields/grid/file_grid,cp/date_picker,cp/files/manager,cp/files/copy-url,cp/files/menu&v=1689006136:2965
(anonymous) @ admin.php?/cp&C=javascript&M=combo_load&ui=droppable,draggable&file=cp/files/picker,fields/file/control_panel,fields/file/file_field_drag_and_drop,cp/publish/entry-list,fields/file/concurrency_queue,fields/file/file_upload_progress_table,fields/file/drag_and_drop_upload,fields/grid/file_grid,cp/date_picker,cp/files/manager,cp/files/copy-url,cp/files/menu&v=1689006136:3064
dispatch @ jquery.js?v=1689006136:2
v.handle @ jquery.js?v=1689006136:2

**Comment by rmaclemale** [comment]
Actually, full disclosure - my test site had been a copy of our main website, but to do this latest test, I did a clean install of EE 7.3.6.  I wanted to rule out any variables related to upgrading an older site.  This is a brand new install.

**Comment by intoeetive** [comment]
Yeah, that's expected in the Console tab - but what's the `Response` for latest request in Network tab (for the request to `filepicker/ajax-upload`)

**Comment by rmaclemale** [comment]
Ah - sorry.  I've only ever used Inspector to look at my own bad css mistakes.  When I click on the filepicker and click Response, it says this:
<pre>string(3) "gd2"
</pre><pre>bool(true)
</pre>

**Comment by intoeetive** [comment]
Thank you. Unfortunately that's how it's expected to be, so everything seems to be set up correctly on this front - yet now working :(
IDK if I'll be able to get Windows Server 2016 setup, but I'll try to test with IIS on Windows 10 and see what comes out...

**Comment by rmaclemale** [comment]
Thanks, we appreciate it.

**Comment by rmaclemale** [comment]
More info.  I have verified that gd is working on the server.  The following script successfully takes an image, rotates it by 45 degrees, and saves the new version to the server using gd commands:
<?php
$img = imagecreatefromjpeg("field.jpg");
$imgRotated = imagerotate($img, 45, +1);
imagejpeg($imgRotated, "myFieldRotated.jpg", 100);
?>
So I'm thinking this is an actual bug.  It's definitely holding us up from upgrading to EE 7, unfortunately.  Believe me, I wish we ran Linux instead of Windows Server, but I have no say in the matter.
